### ISSUE TITLE
MSSQL ERROR: setFirstResult and setMaxResults. with string ID

### ISSUE THREAD

**Issue by pablocarreraest** [typeorm/typeorm#388]
Im using for pagination this query builder for SQL Server:
```javascript
try {
  let posts = await postRepository
                .createQueryBuilder("post")
                .where("post.deletedAt IS NULL")
                .setFirstResult(5)
                .setMaxResults(10)
                .getMany();
  reply(JSON.stringify(posts)).code(200).type('application/json')
} catch (err) {
  reply(JSON.stringify(err)).code(500).type('application/json');
}
```
And I get and error of:
```json
{
  "name": "RequestError",
  "message": "Invalid usage of the option NEXT in the FETCH statement.",
  "code": "EREQUEST",
  "number": 153,
  "lineNumber": 1,
  "state": 2,
  "class": 15,
  "serverName": "SQLServer",
  "procName": "",
  "precedingErrors": [
    {
      "name": "RequestError",
      "message": "Incorrect syntax near 'OFFSET'.",
      "code": "EREQUEST",
      "number": 102,
      "lineNumber": 1,
      "state": 1,
      "class": 15,
      "serverName": "SQLServer",
      "procName": ""
    }
  ]
}
```
For that it seem's to be that SQL Server version 2008 don't support at 
[Invalid usage of the option NEXT in the FETCH statement](http://stackoverflow.com/a/19617772/2116209)

then I try to use limit and offset:
```javascript
try {
  let posts = await postRepository
                .createQueryBuilder("post")
                .where("post.deletedAt IS NULL")
                .offset(5)
                .limit(10)
                .getMany();
  reply(JSON.stringify(posts)).code(200).type('application/json')
} catch (err) {
  reply(JSON.stringify(err)).code(500).type('application/json');
}
```
I get 500 a Internal Server Error without a message.
the  best approach should be updating the mssql?

finally I update to mssql 2016 and found that if you use a primary key type string in this query:
```javascript
let post = await postRepository
                .createQueryBuilder("p")
                .setFirstResult(0)
                .setMaxResults(10)
                .getMany();
```
It will give you a error in the query generated in:
```sql
// this works fine
executing query: 
SELECT DISTINCT("distinctAlias"."u_stringID") as ids_entity 
FROM (
SELECT "u"."stringID" AS "u_stringID",  "u"."name" AS "u_name", "u"."createdAt" AS "u_createdAt", "u"."updatedAt" AS "u_updatedAt", "u"."deletedAt" AS "u_deletedAt" 
FROM "entity" "u") "distinctAlias" 
ORDER BY "ids_entity" 
OFFSET 0 ROWS FETCH NEXT 10 ROWS ONLY
// this fails because strindID# treats as number without '' quotes
[0] executing query:
 SELECT "u"."stringID" AS "u_stringID", "u"."name" AS "u_name", "u"."createdAt" AS "u_createdAt", "u"."updatedAt" AS "u_updatedAt", "u"."deletedAt" AS "u_deletedAt" 
FROM "entity" "u" 
WHERE "u"."stringID" 
IN (stringID1, stringID2, stringID3, stringID4, stringID5, stringID6, stringID7, stringID8, stringID9, stringID10 )
```
It seems it fails because treats as a number, without quotes.
but in my entity class a declare as:
```typescript
@PrimaryColumn('string', {length:'32'})
    stringID: string;
```

**Comment by pablocarreraest** [comment]
Why when I use setFirstResult and setMaxResult use a DISTINCT?
```typescript
let posts = await userRepository
                .createQueryBuilder('p')
                .select(['p.name', 'p.deletedAt'])
                .from(Post,'p')
                .where('p.deletedAt IS NULL')
                .orderBy('p.name','ASC')
                .setFirstResult(0)
                .setMaxResults(10)
                .getMany();
```
DISTINCT is unnecessary for pagination:
```sql
SELECT DISTINCT("distinctAlias"."p_id") as ids_post, "distinctAlias"."p_name" 
FROM (SELECT "p"."name", "p"."deletedAt" FROM "post" "p" WHERE "p"."deletedAt" IS NULL) "distinctAlias" 
ORDER BY "distinctAlias"."p_name" ASC 
OFFSET 0 ROWS FETCH NEXT 10 ROWS ONLY
``` 
Just want to use this query for pagination:
```sql
SELECT "p"."name", "p"."deletedAt" 
FROM "post" "p"
WHERE "p"."deletedAt" IS NULL
ORDER BY "p"."name" ASC 
OFFSET 0 ROWS FETCH NEXT 10 ROWS ONLY;
```
Or something I'm doing wrong

**Comment by pleerock** [comment]
pagination is not a simple process once you get more complicated data returned by a database for you such as joins. It makes two queries to resolve related issues and make pagination really correct. If you don't want this complexity and you simply need to execute:

```sql
SELECT "p"."name", "p"."deletedAt" 
FROM "post" "p"
WHERE "p"."deletedAt" IS NULL
ORDER BY "p"."name" ASC 
OFFSET 0 ROWS FETCH NEXT 10 ROWS ONLY;
```

then simply use `setLimit(...).setOffset(...)` instead of setFirstResult and setMaxResults methods.

**Comment by pablocarreraest** [comment]
@pleerock I try of every way a found in the QueryBuilder class:
with setLimit(10) and setOffset(0)
```typescript
let posts = await postRepository
                .createQueryBuilder('p')
                .select(['p.name', 'p.deletedAt'])
                .where('p.deletedAt IS NULL')
                .setOffset(0)
                .setLimit(10)
                .getMany();
```
generate this:
```sql
SELECT "p"."name", "p"."deletedAt" 
FROM "post" "p" 
WHERE "p"."deletedAt" IS NULL 
LIMIT 10
```
Don't generate the 
```sql 
OFFSET 0 ROWS FETCH NEXT 10 ROWS ONLY;
 ```

**Comment by pleerock** [comment]
if so then its a bug. Basically setFirstResult and setMaxResult should work for you however looks like your issue is because you have `sting id`, not `integer`.

**Comment by pablocarreraest** [comment]
@pleerock working in this issue I found that at method getEntitiesAndRawResults():
after you run the first query and get the ids
```typescript
const ids = results.map(result => result["ids_" + metadata.firstPrimaryColumn.propertyName]);
const areAllNumbers = ids.map((id: any) => typeof id === "number");
// easy solution will be
const areAllStrings = ids.map((id: any) => typeof id === "string");
if (areAllNumbers) {
 // fixes #190. if all numbers then its safe to perform query without parameter
  condition = `${mainAliasName}.${metadata.firstPrimaryColumn.propertyName} IN (${ids.join(", ")})`;
} else if(areAllStrings) {
  // fixes #388. if all numbers then its safe to perform query without parameter I have to test
  condition =`${mainAliasName}.${metadata.firstPrimaryColumn.propertyName} IN ('${ids.join("', '")})`;
} else {
  parameters["ids"] = ids;
  condition = mainAliasName + "." + metadata.firstPrimaryColumn.propertyName + " IN (:ids)";
}
```
Also you can don't separate the queries in 2 and get embed in one. That also work! Im just reviewing

**Comment by pleerock** [comment]
right thats a fix. You could create a PR for this fix.

**Comment by pleerock** [comment]
Fix for this issue is available from `0.1.0-alpha.26`.
